src/main/java/com/ibm/etcd/client/EtcdClient.java

```diff
@@ -338,32 +336,47 @@ public void close() {
             try {
                 channel.shutdown().awaitTermination(2, SECONDS);
             } catch (InterruptedException e) {}
-            executeWhenIdle(() -> internalExecutor.shutdownGracefully(0, 1, SECONDS));
-        });
+            executeWhenIdle(() -> internalExecutor.shutdownGracefully(0, 1, SECONDS), false);
+        }, true);
     }
     
-    // used only in clean shutdown logic
-    final Iterable<SingleThreadEventLoop> eventLoops;
-    
     /**
      * Execute the provided task in the EventLoopGroup only once there
      * are no more running/queued tasks (but might be future scheduled tasks).
      * The key thing here is that it will continue to wait if new tasks
      * are scheduled by the already running/queued ones.
      */
-    private void executeWhenIdle(Runnable task) {
+    private void executeWhenIdle(Runnable task, boolean yield) {
+        AtomicInteger remainingTasks = new AtomicInteger(-1);
+        // Two "cycles" are performed, the first with remainingTasks == -1.
+        // If remainingTasks > 0 after the second cycle, this method
+        // is re-called recursively (in an async context)
         CyclicBarrier cb = new CyclicBarrier(internalExecutor.executorCount(), () -> {
-            if(Iterables.all(eventLoops, ex -> ex.pendingTasks() == 0)) task.run();
-            else executeWhenIdle(task);
+            int rt = remainingTasks.get();
+            if(rt == -1) remainingTasks.incrementAndGet();
+            else if(rt > 0) executeWhenIdle(task, yield);
+            else if(yield) internalExecutor.execute(task);
+            else task.run();
+        });
+        internalExecutor.forEach(ex -> {
+            ex.execute(new Runnable() {
```

Comment:
why not using lambda here `() -> {}`?

Reply:
because the `Runnable` uses the `this` keyword to re-schedule itself, which you can't do with lambdas

Reply:
did not see `this` keyword usage here. I tried locally, it compiled without errors. also, line 361 `internalExecutor.forEach(ex -> {` can be simplified to `internalExecutor.forEach(ex -> `, but I am ok with keeping outside `{}` to make it more readable

Reply:
It's here:
```java
                    if(stel.pendingTasks() > 0) ex.execute(this);
```
I'm surprised that it compiles for you because `this` will then refer to the enclosing `EtcdClient` instance which isn't a `Runnable` (and so can't be passed to the `execute()` method).

But good spot on the braces simplification, i'll push an update with that.

Reply:
oh, I must have left that out. I am ok with this pr now.